examples/two_dimensional_mountain_wave.jl

```diff
@@ -0,0 +1,386 @@
+fig_terrain
+
+# ## Rayleigh damping layer
+#
+# A sponge layer at the top of the domain prevents spurious wave reflections from
+# the upper boundary. The damping follows a Gaussian profile centered at the domain top:
+#
+# ```math
+# \mathcal{S}(\rho w) = -\omega \, \exp\left(-\frac{(z - z_0)^2}{2 \Delta z^2}\right) \rho w
+# ```
+#
+# where ``\omega = 1/60 \, {\rm s}^{-1}`` is the relaxation rate and ``\Delta z = L_z/2``.
+
+@inline gaussian_mask(z, p) = exp(-(z - p.z0)^2 / 2p.dz^2)
+
+@inline function sponge(i, j, k, grid, clock, params, ρc, lz)
+    z = znode(k, grid, lz)
+    ω = params.ω
+    m = gaussian_mask(z, params)
+    return @inbounds -ω * m * ρc[i, j, k]
+end
+
+@inline ρw_sponge(i, j, k, grid, clock, model_fields, params) =
+    sponge(i, j, k, grid, clock, params, model_fields.ρw, Face())
+
+sponge_params = (z0 = grid.Lz, dz = grid.Lz / 2, ω = 1/60)
+ρw_forcing = Forcing(ρw_sponge, discrete_form=true, parameters=sponge_params)
+
+# ## Model initialization
+#
+# Create the atmosphere model with the anelastic formulation, 5th-order WENO advection,
+# and the Rayleigh damping layer:
+reference_state = ReferenceState(grid, constants, surface_pressure=p₀, potential_temperature=θ₀)
+formulation = AnelasticFormulation(reference_state)
+
+advection = WENO(order=5)
+model = AtmosphereModel(grid; formulation, advection, forcing=(; ρw=ρw_forcing))
+
+# ## Initial conditions
+#
+# Initialize with the constant-``N`` stratification and uniform background wind.
+# The potential temperature profile that maintains constant Brunt–Väisälä frequency is:
+#
+# ```math
+# \theta(z) = \theta_0 \exp\left(\frac{N^2 z}{g}\right)
+# ```
+
+θᵢ(x, z) = θ₀ * exp(N² * z / g)
+set!(model, θ = θᵢ, u = U)
+
+# ## Simulation
+#
+# Run for 2 hours with a fixed time step. A small time step is required for numerical stability with this test case:
+
+Δt = 2.0            # s - time step (reduced for stability)
+stop_time = 2hours  # total simulation time
+
+simulation = Simulation(model; Δt, stop_time, align_time_step=false)
+
+# Progress callback to monitor simulation health:
+
+wall_clock = Ref(time_ns())
+
+function progress(sim)
+    elapsed = 1e-9 * (time_ns() - wall_clock[])
+
+    msg = @sprintf("Iter: %d, time: %s, wall time: %s, max|w|: %.3e m s⁻¹",
+                   iteration(sim), prettytime(sim), prettytime(elapsed),
+                   maximum(abs, sim.model.velocities.w))
+
+    wall_clock[] = time_ns()
+
+    @info msg
+
+    return nothing
+end
+
+add_callback!(simulation, progress, name=:progress, IterationInterval(200))
+
+# ## Output
+#
+# Save velocity fields for post-processing and validation:
+
+filename = "mountain_waves"
+simulation.output_writers[:fields] = JLD2Writer(model, model.velocities;
+                                                filename,
+                                                schedule = TimeInterval(100),
+                                                overwrite_existing = true)
+
+run!(simulation)
```

Comment:
Simulation is stopping because there are NaNs: https://numericalearth.github.io/BreezeDocumentation/previews/PR333/literated/two_dimensional_mountain_wave/#Output
```
[ Info: time = 600.0, iteration = 300: NaN found in field ρu. Stopping simulation.
```

Reply:
It’s weird that the simulation blows up in CI but not on my end. 😢

Reply:
What system? Can you show the output of `versioninfo()` in the REPL? You can see at https://numericalearth.github.io/BreezeDocumentation/previews/PR333/literated/two_dimensional_mountain_wave/#Julia-version-and-environment-information the information about the CI machine.

Reply:
The CI system uses an AMD CPU, unlike the two machines I just tested. Could that be the issue?

Reply:
One thing I did notice is that the RNG seems to generate different streams of numbers for different GPUs models/generations, but I don't think we're using random numbers here, right?

Reply:
I could replicate the NaNs with
```julia-repl
julia> versioninfo()
Julia Version 1.12.2
Commit ca9b6662be4 (2025-11-20 16:25 UTC)
Build Info:
  Official https://julialang.org release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 192 × AMD Ryzen Threadripper PRO 7995WX 96-Cores
  WORD_SIZE: 64
  LLVM: libLLVM-18.1.7 (ORCJIT, znver4)
  GC: Built with stock GC
Threads: 1 default, 1 interactive, 1 GC (on 192 virtual cores)
```
with NVIDIA GeForce RTX 5090.

Reply:
> On the same machine I could also replicate the NaN with the CPU architecture (which I presume is what also @navidcy tested), i.e. by replacing `GPU()` at
> 
> https://github.com/NumericalEarth/Breeze.jl/blob/9743747540a39d4db5917a9685ce91a0b0798fc1/examples/two_dimensional_mountain_wave.jl#L145
> 
> to `CPU()`. @kaiyuan-cheng on the machines where you don't have NaNs, what do you get with the CPU simulation? It fails in 80 seconds for me, it doesn't run for long.

This case runs fine on both CPU and GPU on the two machines I tested. Very strange.